src/hooks/useQueryParams.tsx

```diff
@@ -0,0 +1,60 @@
+import { isEqual } from "lodash";
+import queryString from "query-string";
+import { Dispatch, SetStateAction, useEffect, useState } from "react";
+import { useHistory, useLocation } from "react-router-dom";
+
+export interface QueryParams {
+  [key: string]: string | number | undefined;
+}
+
+export type HistoryAction = "replace" | "push";
+
+// A basic hook for fetching query param values and setting query param values
+const useQueryParams = (
+  defaultValues: QueryParams,
+  defaultAction: HistoryAction = "replace",
+): [
+  QueryParams,
+  Dispatch<SetStateAction<QueryParams>>,
+  Dispatch<SetStateAction<HistoryAction>>,
+] => {
+  const history = useHistory();
+  const location = useLocation();
+
+  const defaultQueryParams = queryString.parse(location.search) as QueryParams;
+  const hasExistingURLQueryParams = Object.keys(defaultQueryParams).length > 0;
+
+  const initialValues = hasExistingURLQueryParams
+    ? defaultQueryParams
+    : defaultValues;
+  const [values, setValues] = useState<QueryParams>(initialValues);
+  const [action, setAction] = useState<HistoryAction>(defaultAction);
+
+  // If values have been updated, also set the query params
+  useEffect(() => {
+    const currentQuery = queryString.parse(location.search);
+
+    if (!isEqual(currentQuery, values)) {
+      location.search = queryString.stringify(values);
+      if (action == "replace") {
+        history.replace(location);
+      }
+      if (action == "push") {
+        history.push(location);
+      }
+    }
```

Comment:
Actually, maybe it makes sense to return the following:
- `pushState`
- `replaceState`

Where `replaceState` will get called when the user edits the form values, and `pushState` will get  called when the user clicks copy link.

edit: okay that's actually kind of the same as `createNewState(action, values)`, either approach works

Reply:
I second this, but I don't think it's a showstopper for the initial release. Wouldn't be too hard to go back and change later since we're only planning to use it in one or two places.

Reply:
Those are all good points, and things I considered. Since you are the client of this hook, if that design makes more sense, I'll update it to return two kinds of `setValues`.